src/libponyc/pass/docgen.c

```diff
@@ -755,6 +797,142 @@ static void doc_methods(docgen_t* docgen, docgen_opt_t* docgen_opt,
     doc_method(docgen, docgen_opt, p->ast);
 }
 
+static char* concat(const char *s1, const char *s2)
+{
+  size_t str_size = strlen(s1) + strlen(s2) + 1;
+  char* result = (char*) malloc(str_size); //+1 for the null-terminator
+  strcpy(result, s1);
+  strcat(result, s2);
+  result[str_size - 1] = '\0';
+  return result;
+}
+
+static doc_sources_t* copy_source_to_doc_src(docgen_t* docgen, source_t* source, const char* package_name)
+{
+  doc_sources_t* result = (doc_sources_t*) calloc(sizeof(doc_sources_t), 1);
+
+  char filename_copy[FILENAME_MAX];
+  strcpy(filename_copy, source->file);
+  const char* filename = get_file_name(filename_copy);
+  const char* filename_without_ext = remove_ext(filename, '.', 0);
+  const char* filename_md_extension = concat(filename_without_ext, ".md");
+  char source_dir[FILENAME_MAX];
+  const char* doc_source_dir_relative = concat("src/", package_name);
+  doc_source_dir_relative = concat(doc_source_dir_relative, "/");
+  path_cat(docgen->doc_source_dir, package_name, source_dir);
+
+  pony_mkdir(source_dir);
+  
+  char* path = (char*) malloc(sizeof(char) * FILENAME_MAX);
+  path_cat(source_dir, filename_md_extension, path);
+  const char* doc_path = concat(doc_source_dir_relative, filename_md_extension);
+
+  bool are_name_clashing = false;
+  int current_numbering = 0;
+  int number_of_try = 0;
+  do {
```

Comment:
I am not sure if this part is necessary. Can 2 files in the same package have the same name?